src/Microsoft.AspNet.Authentication/AuthenticationHandler.cs

```diff
@@ -395,11 +235,48 @@ public virtual bool ShouldConvertChallengeToForbidden()
         /// deals an authentication interaction as part of it's request flow. (like adding a response header, or
         /// changing the 401 result to 302 of a login page or external sign-in location.)
         /// </summary>
-        /// <returns></returns>
-        protected virtual Task ApplyResponseChallengeAsync()
+        /// <param name="context"></param>
+        /// <returns>True if no other handlers should be called</returns>
+        protected virtual Task<bool> HandleUnauthorizedAsync(ChallengeContext context)
         {
-            ApplyResponseChallenge();
-            return Task.FromResult(0);
+            Response.StatusCode = 401;
+            return Task.FromResult(false);
+        }
+
+        public async Task ChallengeAsync(ChallengeContext context)
+        {
+            bool handled = false;
+            ChallengeCalled = true;
+            if (ShouldHandleScheme(context.AuthenticationScheme))
+            {
+                switch (context.Behavior)
+                {
+                    case ChallengeBehavior.Automatic:
+                        // If there is a principal already, invoke the forbidden code path
+                        var ticket = await AuthenticateAsync();
+                        if (ticket?.Principal != null)
+                        {
+                            handled = await HandleForbiddenAsync(context);
+                        }
+                        else
+                        {
+                            handled = await HandleUnauthorizedAsync(context);
+                        }
+                        break;
+                    case ChallengeBehavior.Unauthorized:
+                        handled = await HandleUnauthorizedAsync(context);
+                        break;
+                    case ChallengeBehavior.Forbidden:
+                        handled = await HandleForbiddenAsync(context);
+                        break;
+                }
+                context.Accept();
```

Comment:
I think so, they are two different things.  Handled is really don't call anyone else.  If you don't accept, we blow up :/